File: rubygems/pray-cli/lib/pray/cache_clean.rb
# frozen_string_literal: true

require "fileutils"

module Pray
  module CacheClean
    module_function

    def clean_unused_registry_cache(project_root)
      lockfile = Pray.read_lockfile(File.join(project_root, "Prayfile.lock"))
      validate_lockfile_for_cleanup!(lockfile)
      registry_root = File.expand_path(".pray/cache/registry", project_root)
      retained = lockfile.package.filter_map do |package|
        package_path = File.expand_path(package.path, project_root)
        package_path if PathSafety.path_under_root?(registry_root, package_path)
      end

      begin
        metadata = File.lstat(registry_root)
      rescue Errno::ENOENT
        return
      end
      if metadata.symlink? || !metadata.directory?
        remove_path(registry_root)
        return
      end
      prune_directory(registry_root, retained, remove_when_empty: false)
    end

    def validate_lockfile_for_cleanup!(lockfile)
      validate_sha256_digest!("manifest_hash", lockfile.manifest_hash)
      lockfile.package.each do |package|
        if !package.path.is_a?(String) || package.path.empty?
          raise Error.parse("lockfile", "package path must not be empty")
        end
        validate_sha256_digest!("package tree_hash", package.tree_hash)
        validate_sha256_digest!("package artifact_hash", package.artifact_hash)
      end
    end

    def validate_sha256_digest!(field, value)
      digest = value.to_s.delete_prefix("sha256:")
      return if value.is_a?(String) && value.start_with?("sha256:") && digest.match?(/\A[0-9a-f]{64}\z/)

      raise Error.parse("lockfile", "#{field} must be a sha256 digest")
    end

    def prune_directory(path, retained, remove_when_empty:)
      Dir.each_child(path) do |name|
        entry = File.join(path, name)
        protects_entry = retained.include?(entry)
        leads_to_retained = retained.any? { |kept| PathSafety.path_under_root?(entry, kept) }
        next if protects_entry

        metadata = File.lstat(entry)
        if leads_to_retained && metadata.directory? && !metadata.symlink?
          prune_directory(entry, retained, remove_when_empty: true)
        else
          remove_path(entry)
        end
      end
      Dir.rmdir(path) if remove_when_empty && Dir.empty?(path)
    end

    def remove_path(path)
      metadata = File.lstat(path)
      if metadata.directory? && !metadata.symlink?
        FileUtils.rm_rf(path)
      else
        File.unlink(path)
      end
    rescue Errno::ENOENT
      nil
    end
  end
end

File: crates/pray-cli/src/cache_clean.rs
use crate::materialize::remove_path_if_exists;
use pray_core::lockfile::read_lockfile;
use pray_core::{PrayError, PrayResult};
use std::fs;
use std::path::{Component, Path, PathBuf};

pub(crate) fn clean_unused_registry_cache(project_root: &Path) -> PrayResult<()> {
    let lockfile = read_lockfile(&project_root.join("Prayfile.lock"))?;
    validate_lockfile_for_cleanup(&lockfile)?;
    let registry_root = project_root.join(".pray/cache/registry");
    let retained = lockfile
        .package
        .iter()
        .filter_map(|package| retained_registry_path(project_root, &registry_root, &package.path))
        .collect::<Vec<_>>();

    let metadata = match fs::symlink_metadata(&registry_root) {
        Ok(metadata) => metadata,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(error) => return Err(error.into()),
    };
    if metadata.file_type().is_symlink() || !metadata.is_dir() {
        return remove_path_if_exists(&registry_root);
    }

    prune_directory(&registry_root, &retained, false)
}

fn validate_lockfile_for_cleanup(lockfile: &pray_core::lockfile::Lockfile) -> PrayResult<()> {
    validate_sha256_digest("manifest_hash", &lockfile.manifest_hash)?;
    for package in &lockfile.package {
        if package.path.is_empty() {
            return Err(lockfile_parse_error("package path must not be empty"));
        }
        validate_sha256_digest("package tree_hash", &package.tree_hash)?;
        validate_sha256_digest("package artifact_hash", &package.artifact_hash)?;
    }
    Ok(())
}

fn validate_sha256_digest(field: &str, value: &str) -> PrayResult<()> {
    let digest = value.strip_prefix("sha256:").unwrap_or_default();
    if digest.len() == 64
        && digest
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
    {
        Ok(())
    } else {
        Err(lockfile_parse_error(&format!(
            "{field} must be a sha256 digest"
        )))
    }
}

fn lockfile_parse_error(message: &str) -> PrayError {
    PrayError::Parse {
        kind: "lockfile",
        message: message.to_string(),
    }
}

fn retained_registry_path(
    project_root: &Path,
    registry_root: &Path,
    package_path: &str,
) -> Option<PathBuf> {
    let candidate = Path::new(package_path);
    let absolute = if candidate.is_absolute() {
        lexical_normalize(candidate)
    } else {
        lexical_normalize(&project_root.join(candidate))
    };
    absolute
        .strip_prefix(lexical_normalize(registry_root))
        .ok()
        .map(|relative| registry_root.join(relative))
}

fn lexical_normalize(path: &Path) -> PathBuf {
    let mut normalized = PathBuf::new();
    for component in path.components() {
        match component {
            Component::CurDir => {}
            Component::ParentDir => {
                normalized.pop();
            }
            other => normalized.push(other.as_os_str()),
        }
    }
    normalized
}

fn prune_directory(path: &Path, retained: &[PathBuf], remove_when_empty: bool) -> PrayResult<()> {
    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let entry_path = entry.path();
        let protects_entry = retained.iter().any(|kept| kept == &entry_path);
        let leads_to_retained = retained.iter().any(|kept| kept.starts_with(&entry_path));
        if protects_entry {
            continue;
        }

        let metadata = fs::symlink_metadata(&entry_path)?;
        if leads_to_retained && metadata.is_dir() && !metadata.file_type().is_symlink() {
            prune_directory(&entry_path, retained, true)?;
        } else {
            remove_path_if_exists(&entry_path)?;
        }
    }

    if remove_when_empty && fs::read_dir(path)?.next().is_none() {
        fs::remove_dir(path)?;
    }
    Ok(())
}
